Allow building stage1+ std with `panic=abort`

clubby789 opened the pull request:
Initial approach to #84766

This adds a `rust.panic-strategy-std` option to bootstrap, which controls how stage1+ std is built. I decided not to apply it to stage 0 std, otherwise I'd need to make the bootstrap compiler inject the correct panic runtime to build stage1 correctly.

Draft for now as this I'm not sure about the strategy, and only stage0 and 1 can be built with this option right now.

(All library changes are just fixing warnings that occur when building std with `panic_immediate_abort`)

saethlin commented:
I'm very wary of the proliferation and complexity of the `cfg(feature = "panic_immediate_abort")` in the library. I don't really have any confidence in our ability to keep the feature working because it has no tests: https://github.com/rust-lang/rust/issues/118393 I tried to follow my own directions in there to create a test and I got too frustrated by the complexity of the run-make tests which I think is what one would use here, and I gave up.

saethlin commented on src/bootstrap/src/lib.rs:
```diff
@@ -707,8 +708,12 @@ impl Build {
-        let mut features = " panic-unwind".to_string();
+    fn std_features(&self, target: TargetSelection, stage: u32) -> String {
+        let mut features = match self.config.rust_panic_strategy_std {
+            PanicStrategy::Abort if stage > 0 => " panic_immediate_abort",
```
I do not think that we should be doing this, aborting panics are not the same as `panic_immediate_abort`; aborting panics are effectively just panics that don't unwind the stack. They still do all the formatting of the panic message and print a backtrace. Immediate-abort panics don't do any formatting and just call `core::intrinsics::abort.

I have thought about ways to make immediate-abort a proper panic strategy, and one could do that albiet poorly by changing the lowering of calls to the panic lang items when the feature is enabled. That doesn't fix all the `cfg`s which are most of the maintenance burden.

saethlin commented:
I'm very happy to see someone working on making our support for panic modes other than unwind better; I don't want to stifle making progress by advising too much caution but I do think the fact that we default to unwind and assume unwinding in so many places has laid a lot of traps that it will take time to fix.

clubby789 committed:
Add `panic-strategy` to bootstrap config

clubby789 commented:
Given that if std uses `panic=abort`, anything depending on it will also need that, so I've moved the flag to be set for all Cargo invocations (and removed the `panic_immediate_abort`) stuff.
To get testing to work, I guess for a start I'll need to also set `-Zpanic_abort_tests` for test crates, and pass the panic configuration through to compiletest so it can build with `panic=abort` and ignore tests requiring unwinding.

albertlarsan68 commented:
Is this PR waiting on my review, or is more implementation/bug hunting needed?

clubby789 commented:
More implementation needed - I need to figure out an approach for the above

bors commented:
:umbrella: The latest upstream changes (presumably #120803) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@clubby789 any updates on this?

Dylan-DPC commented:
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
